### PR TITLE
allow calendar service to be called from other services - will close

### PR DESCRIPTION
**Pull request by sdetweil**
####  Description
make calendar service re-entrant and callable from outside services.
move event array outside service implementation

####  Fixes Related issues
- add related
- issues here

####  Checklist
- [x] I have read and understand the CONTRIBUTIONS.md file
- [x] I have searched for and linked related issues
- [x] I have added config.schema.json file if config option are required.
- [x] I am NOT targeting master branch

### CODE REVIEW

**Comment by sdetweil** [comment]
also addresses issue/enhancement #536

**Comment by sdetweil** [comment]
updated to external motion detector to send start/end, and use that to signal like the original.

**Comment by evancohen** [comment]
@justbill2020 have you had the chance to look at this yet? I broke my Pi (it has taken a lot of abuse recently), and while I wait for a new one I'm not going to be able to test GPIO stuff.

**Comment by justbill2020** [comment]
@evancohen... I've done little to nothing the merger just went through and after my vacation last week I'm playing catchup

File: plugins/autosleep/service.js [hunk 1]
```diff
@@ -2,52 +2,96 @@
...
+						EneryStarTimer = null;
+					}
+					// if the dummy wake up delay is running, stop it too
+					if(EneryStarTimerStop !=null){
+						$interval.cancel(EneryStarTimerStop)
+					}
+				}
 			}
 			Focus.change("default");
+			SpeechService.cleanup();
```
Comment (justbill2020) [inline plugins/autosleep/service.js]: what is this doing? I don't see cleanup function in the SpeechService.

**Comment by justbill2020** [comment]
how is the external motion actually working? i see [extwakeup](https://github.com/evancohen/smart-mirror/pull/591/files#diff-fd2ccc258050d751f3ddafc8c39551a4) but when and how is that running? are there additional dependencies that must be installed?

File: plugins/speech/service.js [hunk 2]
```diff
@@ -1,99 +1,104 @@
...
+
+            ipcRenderer.on('final-results', (event, text) => {
+                callbacks.finalResult(text)
+                annyang.trigger(text)
+                callbacks.listening(false)
+            })
+
+        }
+
+				service.cleanup = function(){
```
Comment (justbill2020) [inline plugins/speech/service.js]: ok found the clean up. i think setting an timeout when listening starts... for say 2 minutes to run callbacks.listening(false)

File: plugins/speech/service.js [hunk 3]
```diff
@@ -1,99 +1,104 @@
...
+            }
+            if (isCallback(cb.error)) {
+                callbacks.error = function (data) {
+                    $rootScope.$apply(cb.error(data))
+                }
+            }
+
+            ipcRenderer.on('hotword', () => {
+                callbacks.listening(true)
+            })
```
Comment (justbill2020) [inline plugins/speech/service.js]: add the timeout between line 41 and 42

```javascript
service.clear = setTimeout( () => {
   callbacks.listening(false)
}, 60000 * 2)
```

Comment (justbill2020) [inline plugins/speech/service.js]: should also remove the cleanup call from the autotimer/motion service

**Comment by sdetweil** [comment]
External Motion,  (it does not specify HOW the file gets written to the detector location)

one example: (using a webcam as the motion source)

install the github motion deamon, https://github.com/sackmotion/motion. and configure its external event commands to run (on event start)
extwakeup detected 
and (on event end)
extwakeup ended

the parms matter, as that is the filename the handler (in motion.js)  looks for.

File: plugins/autosleep/config.schema.json [hunk 4]
```diff
@@ -1,157 +1,163 @@
...
+                    "title": "Command used put Smart Mirror to sleep",
+                    "default": "sudo ./scripts/raspi-monitor.sh off > /dev/null 2>&1"
+                }
+            }
+        },
+        "motion": {
+            "type": "object",
+            "title": "Motion Settings",
+            "properties": {
+                "enabled": {
```
Comment (sdetweil) [inline plugins/autosleep/config.schema.json]: ok, I don't see how to do this.. the UI is not well defined for me..

File: plugins/speech/service.js (hunk 2 above)
Comment (sdetweil) [inline plugins/speech/service.js]: the problem with timeout, is that it is just the UI..  we have an issue, #592 which has the UI and the detached sonus task out of synch.   the reset should REALLY get the voice listener reset too.. just don't know how to do that.. (kill and restart?), so the reset doesn't really work..

File: plugins/speech/service.js (hunk 3 above)
Comment (sdetweil) [inline plugins/speech/service.js]: yeh, I know, but I don't think that is the right was, as per my prior comment.. this really has nothing to do with the external motion handler

**Comment by justbill2020** [comment]
so you used this guide?

http://lavrsen.dk/foswiki/bin/view/Motion/MotionGuideInstallation

can you also put the script  into the scripts folder rather than the plugin directory?

**Comment by sdetweil** [comment]
yes, on the motion config
doc page says 
'Motion can output either image files of the picture frames containing motion or movie files showing the entire event. **Motion can also invoke other programs when motion is detected.**

but note again that I don't care HOW you detect the motion.. (this is just an example of the 'how' part) the HOW and the WHAT are separated.. the SM code needs the 'what' (file in folder).. it doesn't KNOW or CARE how it got there (all external to SM code). 
so, you can use some other mechanism, a web server with an api, the hardware pin thing, some other package than 'motion', etc.. 

script.. will do, also added a bunch of setup info to it, and made it a little more resilient.

**Comment by justbill2020** [comment]
I get what you're saying but this isn't just a project for coders... we have to make it available to all kinds of people with different levels of experience and knowledge. If someone is wanting to get a smart-mirror up and running, with "external" motion we must worry about the how... and we have to give everyone the documentation to get it working as well...

**Comment by sdetweil** [comment]
I understand.. I don't see a place in the doc for plugins to add documentation.. else I would have added it

**Comment by justbill2020** [comment]
cool... if you just add the docs for it as an .md file in the plugin folder i can pull it from there and add it to the docs... i'm in the middle of revising them so they are more in a tutorial workflow for specific Operating Systems

**Comment by sdetweil** [comment]
docs ok.. will do.. md file , is there some special format? never written one before.. (looks like text so just asking)

**Comment by justbill2020** [comment]
Take a look at the readme.md file for markdown syntax

**Comment by evancohen** [comment]
This PR contains a whole lot more than just a patch for the Calendar service. Can you either:

1) Remove the new changes so this PR only contains calendar changes
2) Close this and create a new PR with just the changes to the calendar service

**Pull request closed without merging**